### PR TITLE
Make XDP Nodeport Acceleration compatible with EgressGW

### LINKED ISSUE #19624: Add support to bpf_xdp for redirecting packets to a tunnel device
Currently it's not possible to redirect a packet to a tunnel device from the `bpf_xdp` program as there isn't any XDP equivalent helper for `skb_set_tunnel_key()`.

Support for this feature can be added by implementing the logic to manually encapsulate a frame in a VXLAN/Geneve packet with the help of `ctx_adjust_hroom()`, similarly to what we are already doing for [IPIP encap](https://github.com/cilium/cilium/blob/140e82aceebd3be97d6db85d56be729b0af11da1/bpf/lib/nodeport.h#L279-L280) and then doing a FIB lookup to figure out the correct physical interface where the packet should be redirected to.

This will allow to support:
* LB acceleration + tunneling
* XDP acceleration for the return path of egress gateway

### LINKED ISSUE #19717: Egress gateway partially incompatible with LB XDP acceleration
The egress gateway feature is partially incompatible with LB XDP acceleration: tunneling redirection is currently not supported in `bpf_xdp` and so the egress gateway return traffic has to go through the host network stack of the gateway node, which will then route it to the tunnel device.

The issue with that is that the default `FORWARD` iptables chain policy is usually set to `DROP` by docker, resulting in return egress gateway traffic being dropped unless the user explicitly sets it to `ALLOW`.

Once tunneling support for XDP is implemented (https://github.com/cilium/cilium/issues/19624) we can redirect egress gateway traffic from `bpf_xdp` without having to worry about the iptables configuration of the gateway node.

### PR DESCRIPTION
**Pull request by julianwiedmann**
When combining XDP-based Nodeport Acceleration (-> Direct-routing) and Egress Gateway, the XDP program can't redirect the EgressGW reply traffic straight to the `cilium_vxlan` interface. So reply traffic that needs to be forwarded to another node gets passed up to the kernel, where it needs to pass through iptables (and is potentially dropped) before being routed to `cilium_vxlan`.

This PR introduces a fallback mechanism, where the XDP program adds a `needs-encap` flag (+ relevant tunnel info) to the packet before it gets passed up to the kernel. We pick up this flag in the `from-netdev` program @ `TC-Ingress`, and perform the redirect to `cilium_vxlan` from there.

This currently only gets enabled for EgressGW reply traffic, but could get extended to also cover `TUNNEL_MODE` (so that Nodeport-Acceleration can be used without direct-routing). We already lay some groundwork for this.

A second add-on would be to add support for building the encap headers into the XDP program itself. In that case this fallback mechanism would still be taken if eg. the FIB lookup fails, or we only want to use `XDP_TX` (and defer any case where we would need `XDP_REDIRECT`).

Fixes: https://github.com/cilium/cilium/issues/19717
Fixes: https://github.com/cilium/cilium/issues/19624

```release-note
When combining XDP Nodeport Acceleration with Egress Gateway, forwarding the EgressGW reply traffic no longer requires a specific iptables configuration on the Gateway node.
```

### CODE REVIEW

**Review by nebril — APPROVED** [review]
Test bits look good to me.

**Comment by julianwiedmann** [comment]
[rebase to pick up the reshuffled test matrix]

**Comment by julianwiedmann** [comment]
/test-1.16-4.9

**Review by NikAleksandrov — APPROVED** [review]

**Review by NikAleksandrov — APPROVED** [review]
Overall looks good to me, the only 2 minor nits I have are the XFER_FLAGS naming because it looks like they are mutually exclusive and the hardcoded offset validation in ctx_move_xfer(), but these are not a big deal.

**Comment by julianwiedmann** [comment]
> Overall looks good to me, the only 2 minor nits I have are the XFER_FLAGS naming because it looks like they are mutually exclusive and the hardcoded offset validation in ctx_move_xfer(), but these are not a big deal.

ack, we'll need a proper design once we start adding more xfer flags/data. Also a smarter way of doing `ctx_move_xfer()`, instead of sprinkling it over paths that "might" run in XDP context. But right now I didn't want to over-engineer it.

File: bpf/lib/nodeport.h [hunk 1]
```diff
@@ -1006,13 +1001,13 @@ static __always_inline int rev_nodeport_lb6(struct __ctx_buff *ctx, __u32 *ifind
...
 
-	return CTX_ACT_OK;
+	return CTX_ACT_REDIRECT;
 }
 
 __section_tail(CILIUM_MAP_CALLS, CILIUM_CALL_IPV6_NODEPORT_REVNAT)
 int tail_rev_nodeport_lb6(struct __ctx_buff *ctx)
 {
-	int ext_err = 0, ret = 0;
+	int ext_err = 0, ret = 0, ret2 = 0;
```
Comment (qmonnet) [inline bpf/lib/nodeport.h]: Nit: `ret2` could be defined in the `if { }` block, and doesn't have to be initialised

File: bpf/lib/overloadable_xdp.h [hunk 2]
```diff
@@ -106,6 +106,7 @@ ctx_set_encap_info(struct xdp_md *ctx __maybe_unused,
 		   __u32 vni __maybe_unused, __u32 *ifindex __maybe_unused)
 {
 	return DROP_INVALID;
+	/* return CTX_ACT_OK; */
```
Comment (qmonnet) [inline bpf/lib/overloadable_xdp.h]: Leftover? Although I see you remove it in a later commit

File: bpf/lib/common.h [hunk 3]
```diff
@@ -65,6 +65,11 @@
 /* XDP to SKB transferred meta data. */
 #define XFER_PKT_NO_SVC		1 /* Skip upper service handling. */
 
+/* For use in ctx_get_xfer(), after XDP called ctx_move_xfer(). */
+enum {
+	XFER_FLAGS,		/* XFER_PKT_* */
```
Comment (qmonnet) [inline bpf/lib/common.h]: Maybe set explicitly to 0?

File: bpf/lib/common.h [hunk 4]
```diff
@@ -46,6 +46,15 @@
 #define CONDITIONAL_PREALLOC BPF_F_NO_PREALLOC
 #endif
 
+#if defined(ENCAP_IFINDEX) || defined(ENABLE_EGRESS_GATEWAY)
```
Comment (jibi) [inline bpf/lib/common.h:49]: We might be able to simply this a bit (get rid of the `defined(ENABLE_EGRESS_GATEWAY)` check) but let's keep it as part of https://github.com/cilium/cilium/issues/19786 once I start tackling it

**Review by qmonnet — APPROVED** [review]
Looks good! Just a few minor comments below.

File: bpf/lib/overloadable_xdp.h (hunk 2 above)
Comment (julianwiedmann) [reply bpf/lib/overloadable_xdp.h]: I temporarily added this in the initial refactor patch, to motivate where the `CTX_ACT_OK` will be returned from later.

**Comment by julianwiedmann** [comment]
> Looks good! Just a few minor comments below.

Ta! Will pick up those small changes when resolving the `needs-rebase`.

File: bpf/lib/overloadable_xdp.h [hunk 5]
```diff
@@ -123,8 +138,17 @@ ctx_set_encap_info(struct xdp_md *ctx __maybe_unused,
 		   __u32 dstid __maybe_unused,
 		   __u32 vni __maybe_unused, __u32 *ifindex __maybe_unused)
 {
+/* needs more work in the callers: */
+#ifdef TUNNEL_MODE
 	return DROP_INVALID;
-	/* return CTX_ACT_OK; */
+#else
+	ctx_store_meta(ctx, CB_ENCAP_NODEID, bpf_ntohl(node_id));
```
Comment (jibi) [inline bpf/lib/overloadable_xdp.h:145]: Nit, feel free to ignore it.

I'm finding this asymmetry:
* XDP: set these properties with `ctx_store_meta()` beside `XFER_PKT_ENCAP`, which is set with `ctx_set_xfer()`
* TC: read all these properties with `ctx_get_xfer()`

a bit confusing, maybe we can switch all the xfer properties to `ctx_set_xfer()` here (and have an additional helper to set the `XFER_MARKER`)?

**Review by jibi — APPROVED** [review]
Looks good! :rocket: :rocket:

File: bpf/lib/common.h (hunk 4 above)
Comment (julianwiedmann) [reply bpf/lib/common.h:49]: I was hoping for that, yes :+1:. Passing through `TunnelExists()` to BPF.

File: bpf/lib/overloadable_xdp.h (hunk 5 above)
Comment (julianwiedmann) [reply bpf/lib/overloadable_xdp.h:145]: Heh, for me it's actually `ctx_set_xfer()` in XDP what's causing the confusion. As it's not touching the meta_data area, that feels misplaced and should imho be a plain `ctx_store_meta(ctx, XFER_MARKER, ...)` (without the wrapping).

**Review by borkmann — APPROVED** [review]

**Comment by julianwiedmann** [comment]
[Test suite passed previously, and the rebase was trivial. But let's make sure of it ...]

**Comment by julianwiedmann** [comment]
Caught the `test-1.16-4.9` failure that's already been fixed in https://github.com/cilium/cilium/pull/21207. Otherwise all good.

**Pull request merged**